### ISSUE TITLE
Feature request: Step grouping

### ISSUE THREAD

**Issue by Zash** [buildbot/buildbot#6092]
**TL;DR:** A way to group steps, with sensible skipping and failure behavior.

**Background**

Consider a test that runs the same things in multiple configurations. This could be different runtime versions (think Python 2 vs 3.n vs 3.n+1 etc) or different flags to `./configure`.

This could be done by creating a lot of different builds using multiplicative `for` loops, but this might grow out of hand.

Another option might be to simply add multiple sequences of `Configure(), Compile(), Test()` to the same build after an initial VCS checkout, but this makes it difficult to e.g. skip the `Test()` step if `Compile()` fails, as it may be desirable to test Python 3.x+1 even if the Python 3.x test fails.

The `ShellSequence()` step is pretty close to what would be optimal here, but it is limited to shell commands.

**Summary**

So it would be great to have a generic step-grouping step, that respects `haltOnFailure` etc, so that steps within the group would may only halt the group, but the overall build continues.

**Code example**
Usage might be something like this:

```python
build_factory = util.BuildFactory(steps=[steps.FooVCS()])

for foover in ["2.7", "3.7","3.8"]:
    build_factory.addStep(
        steps.Group(
            haltOnFailure=False,
            flunkOnFailure=True,
            steps=[
                steps.Configure(
                    haltOnFailure=True,
                    command=[
                        "./configure",
                        util.Interpolate("--foo-version=%(prop:fooversion)s"),
                    ],
                ),
                steps.Compile(haltOnFailure=True),
                steps.Test(flunkOnFailure=True),
            ],
        )
    )
```

**Comment by p12tic** [comment]
@Zash Why wouldn't you just run multiple builds in this case? Is the source checkout too expensive compared to the build steps?

There could be a "master" build which when started triggers the "worker" builds and waits for their completion using the `Trigger` step.

I understand that a large number of builders would be a problem in the current web UI, but this is a UI issue.

**Comment by Zash** [comment]
We do currently run multiple builds, one for each OS/distro and Lua version, and it works decently after adding a lock to keep the number of concurrent builds down. Our Mercurial repo struggled with around two dozen concurrent clones/pulls and would occasionally time out and cause cause builds to fail.

We should make another attempt at connecting all these with Triggers, that probably gives the best reporting and better scheduling.

We're still figuring out how to arrange things. Feel free to consider this a solution looking for a problem.

**Comment by p12tic** [comment]
@Zash You could add locks for all mercurial steps that clone from the same mercurial repository. This way you would be able to precisely control how many concurrent clones there are.
